File: Behavioral/Observer/Observer.py
from abc import ABC, abstractmethod
from typing import List


class IListener(ABC):

    @abstractmethod
    def update(self):
        pass


class IPublisher(ABC):
    listeners: IListener = []

    @abstractmethod
    def subscribe(self, listener: IListener):
        pass

    @abstractmethod
    def un_subscribe(self, listener: IListener):
        pass

    @abstractmethod
    def notify(self):
        pass


class Publisher(IPublisher):
    listeners: List[IListener] = []

    def subscribe(self, listener: IListener):
        self.listeners.append(listener)

    def un_subscribe(self, listener: IListener):
        self.listeners.pop(listener)

    def notify(self):
        for listener in self.listeners:
            listener.update()


class LoggingListener(IListener):

    def __init__(self, message):
        self.message = message

    def update(self):
        print(f"Logging {self.message}")


class EmailAlertListener(IListener):

    def __init__(self, message):
        self.message = message

    def update(self):
        print(f"Sending email {self.message}")


class EditorEvent:
    publisher: IPublisher

    def __init__(self):
        self.publisher = Publisher()

    def open_file(self):
        self.publisher.notify()


editor = EditorEvent()
logger = LoggingListener("hello")
email = EmailAlertListener("hello")

editor.publisher.subscribe(logger)
editor.publisher.subscribe(email)


editor.open_file()

File: Observer/Observer.go
package main

import "fmt"

type IListener interface {
	update()
}

type IPublisher interface {
	notify()
	subscribe(listener IListener)
	unSubscribe(listener IListener)
}

// Publisher.
type Publisher struct {
	listerners []IListener
}

func newPublisher() IPublisher {
	return &Publisher{listerners: make([]IListener, 0)}
}

func (p *Publisher) subscribe(listener IListener) {
	p.listerners = append(p.listerners, listener)

}

func (p *Publisher) unSubscribe(listener IListener) {
	p.listerners = p.listerners[:len(p.listerners)-1]
}

func (p *Publisher) notify() {
	for _, listener := range p.listerners {
		listener.update()
	}
}

// Logging listener.

type LoggingListener struct {
	message string
}

func newLoggingListener(message string) IListener {
	return &LoggingListener{
		message: message,
	}
}

func (ll *LoggingListener) update() {
	fmt.Printf("Logging listener: %s \n", ll.message)
}

// Email Alert.
type EmailAlert struct {
	message string
}

func newEmailAlert(message string) IListener {
	return &EmailAlert{
		message: message,
	}
}

func (ll *EmailAlert) update() {
	fmt.Printf("Logging listener: %s \n", ll.message)
}

// Event.
type Event struct {
	publisher IPublisher
}

func newEvent(publisher IPublisher) *Event {
	return &Event{
		publisher: publisher,
	}
}

func (e *Event) openFile() {
	e.publisher.notify()
}

func main() {
	publisher := newPublisher()
	event := newEvent(publisher)
	logger := newLoggingListener("hello")
	email := newEmailAlert("hello")

	event.publisher.subscribe(logger)
	event.publisher.subscribe(email)

	event.openFile()
}
